File: components/sound.py
import subprocess, re
from PyQt6.QtWidgets import QLabel, QWidget, QComboBox, QSlider, QPushButton, QMessageBox
from PyQt6.QtGui import QFont
from PyQt6.QtCore import Qt, QTimer


class SoundManager:
    def __init__(self, parent):
        self.parent = parent
        self.audio_timer, self.mic_timer = None, None
        self.refresh_timer = QTimer()
        self.refresh_timer.timeout.connect(self.refresh_state)
        self.playback_sliders = []
        self.playback_cache = {}
        self.init_ui()

    def run_cmd(self, cmd):
        try:
            result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=10)
            return result.stdout.strip() if result.returncode == 0 else ""
        except Exception as e:
            print(f"Command error: {e}")
            return ""

    def dropdown_style(self):
        return """
            QComboBox {
                background: rgba(255, 255, 255, 0.2);
                color: white;
                border: 1px solid rgba(255, 255, 255, 0.2);
                border-radius: 36px;
                padding: 5px;
                font-size: 16px;
            }
            QComboBox QAbstractItemView {
                background: rgba(0, 0, 0, 0.2);
                border: 1px solid rgba(255, 255, 255, 0.2);
                color: black;
                padding: 5px;
                margin: 0px;
                selection-background-color: transparent;
            }
        """

    def slider_style(self):
        return """
            QSlider::horizontal {
                background: transparent;
                border: none;
                padding: 0px;
                margin: 0px;
            }
            QSlider::groove:horizontal {
                height: 8px;
                background: transparent;
                border: none;
                padding: 0px;
                margin: 0px;
            }
            QSlider::handle:horizontal {
                background: pink;
                border: 1px solid white;
                width: 18px;
                margin: -6px 0;
                border-radius: 9px;
            }
            QSlider::sub-page:horizontal {
                background: rgba(255, 255, 255, 0.25);
                border-radius: 4px;
            }
        """

    def init_ui(self):
        self.sound_box = QWidget(self.parent)
        self.sound_box.setGeometry(500, 100, 600, 600)
        self.sound_box.setStyleSheet("background-color: rgba(255,255,255,0.05); border: 2px solid rgba(255,255,255,0.15); border-radius: 30px;")
        self.sound_box.setVisible(False)

        font = QFont("ElysiaOSNew", 14)

        def add_label(text, y):
            label = QLabel(text, self.sound_box)
            label.setFont(font)
            label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            label.move(220, y)
            return label

        def add_dropdown(y, callback):
            dropdown = QComboBox(self.sound_box)
            dropdown.setGeometry(150, y, 300, 40)
            dropdown.setStyleSheet(self.dropdown_style())
            dropdown.currentIndexChanged.connect(callback)
            return dropdown

        def add_slider(y, change_cb):
            slider = QSlider(Qt.Orientation.Horizontal, self.sound_box)
            slider.setGeometry(100, y, 400, 30)
            slider.setMinimum(0)
            slider.setMaximum(100)
            slider.setStyleSheet(self.slider_style())
            slider.valueChanged.connect(change_cb)
            return slider

        self.audio_label = add_label("Audio Output", 30)
        self.audio_dropdown = add_dropdown(60, self.set_audio_output)
        self.audio_slider = add_slider(110, self.debounce_audio)

        self.mic_label = add_label("Microphone Input", 170)
        self.mic_dropdown = add_dropdown(200, self.set_mic_input)
        self.mic_slider = add_slider(250, self.debounce_mic)

        self.apply_btn = QPushButton("Apply Changes", self.sound_box)
        self.apply_btn.setGeometry(200, 510, 200, 50)
        self.apply_btn.setStyleSheet("""
            QPushButton {
                background-color: #e5a7c6;
                color: white;
                font-weight: bold;
                border: none;
                padding: 6px 12px;
                letter-spacing: 1px;
                border-radius: 10px;
                font-size: 16px;
            }
            QPushButton:hover {
                background-color: #edcee3;
            }
        """)
        self.apply_btn.clicked.connect(self.apply_changes)

    def get_ports_for_device(self, name, type_):
        lines = self.run_cmd(f"pactl list {type_}s").splitlines()
        ports, in_block, in_ports = [], False, False
        for line in lines:
            if f"Name: {name}" in line:
                in_block = True
            elif in_block and f"{type_.capitalize()} #" in line:
                break
            elif in_block and "Ports:" in line:
                in_ports = True
            elif in_ports and line.strip() == "":
                in_ports = False
            elif in_ports and ':' in line:
                match = re.match(r'\s*(\S+):\s*([^(]+)', line)
                if match:
                    ports.append((match.group(1), match.group(2).strip()))
        return ports

    def simplify_description(self, text):
        text = text.replace("Analog Stereo", "")
        text = re.sub(r'Family\s+[\w/]+?\s+HD\s+Audio\s+Controller', '', text, flags=re.IGNORECASE)
        text = re.sub(r'[\s\-]+$', '', text)
        return text.strip()

    def get_devices(self, type_):
        short = self.run_cmd(f"pactl list short {type_}s")
        full = self.run_cmd(f"pactl list {type_}s")
        devices = []
        for line in short.splitlines():
            parts = line.split("\t")
            if len(parts) < 2: continue
            dev_id, dev_name = parts[0], parts[1]
            desc = dev_name
            for block in full.split(f"{type_.capitalize()} #")[1:]:
                if f"Name: {dev_name}" in block:
                    for l in block.splitlines():
                        if "Description:" in l:
                            raw = l.split("Description:")[1].strip()
                            desc = self.simplify_description(raw)
                            break
            ports = self.get_ports_for_device(dev_name, type_)
            active = self.get_active_port(dev_name, type_)
            devices.append({"id": dev_id, "name": dev_name, "description": desc, "ports": ports, "active": active})
        return devices

    def get_active_port(self, dev_name, type_):
        lines = self.run_cmd(f"pactl list {type_}s").splitlines()
        found = False
        for line in lines:
            if f"Name: {dev_name}" in line:
                found = True
            elif found and "Active Port:" in line:
                return line.split(":")[1].strip()
            elif found and f"{type_.capitalize()} #" in line:
                break
        return None

    def set_audio_output(self):
        item = self.audio_dropdown.currentData()
        if item:
            device, port = item
            self.run_cmd(f"pactl set-default-sink '{device}'")
            self.run_cmd(f"pactl set-sink-port '{device}' '{port}'")

    def set_mic_input(self):
        item = self.mic_dropdown.currentData()
        if item:
            device, port = item
            self.run_cmd(f"pactl set-default-source '{device}'")
            self.run_cmd(f"pactl set-source-port '{device}' '{port}'")

    def set_audio_volume(self, val):
        item = self.audio_dropdown.currentData()
        if item:
            device, _ = item
            self.run_cmd(f"pactl set-sink-volume '{device}' {val}%")

    def set_mic_volume(self, val):
        item = self.mic_dropdown.currentData()
        if item:
            device, _ = item
            self.run_cmd(f"pactl set-source-volume '{device}' {val}%")

    def debounce_audio(self, val):
        if self.audio_timer: self.audio_timer.stop()
        self.audio_timer = QTimer(singleShot=True, timeout=lambda: self.set_audio_volume(val))
        self.audio_timer.start(100)

    def debounce_mic(self, val):
        if self.mic_timer: self.mic_timer.stop()
        self.mic_timer = QTimer(singleShot=True, timeout=lambda: self.set_mic_volume(val))
        self.mic_timer.start(100)

    def apply_changes(self):
        try:
            self.set_audio_output()
            self.set_mic_input()
            self.set_audio_volume(self.audio_slider.value())
            self.set_mic_volume(self.mic_slider.value())
            QMessageBox.information(self.parent, "Audio", "Settings applied.")
        except Exception as e:
            QMessageBox.warning(self.parent, "Error", f"Failed to apply: {str(e)}")

    def refresh_state(self):
        self.update_slider(self.audio_dropdown, self.audio_slider, is_sink=True)
        self.update_slider(self.mic_dropdown, self.mic_slider, is_sink=False)
        self.refresh_playback_sliders()

    def update_slider(self, dropdown, slider, is_sink):
        item = dropdown.currentData()
        if not item: return
        device, _ = item
        kind = "sink" if is_sink else "source"
        out = self.run_cmd(f"pactl get-{kind}-volume '{device}'")
        match = re.search(r'(\d+)%', out)
        if match:
            vol = min(int(match.group(1)), 100)
            if slider.value() != vol:
                slider.blockSignals(True)
                slider.setValue(vol)
                slider.blockSignals(False)

    def get_playback_clients(self):
        output = self.run_cmd("pactl list sink-inputs")
        clients, current = [], {}
        for line in output.splitlines():
            line = line.strip()
            if line.startswith("Sink Input #"):
                if current: clients.append(current)
                current = {"index": line.split("#")[1]}
            elif line.startswith("application.name"):
                match = re.search(r'"([^"]+)"', line)
                if match: current["name"] = match.group(1)
            elif line.startswith("Volume:"):
                match = re.findall(r'(\d+)%', line)
                if match:
                    volumes = list(map(int, match))
                    avg_volume = sum(volumes) // len(volumes)
                    current["volume"] = avg_volume
        if current: clients.append(current)
        return clients

    def refresh_playback_sliders(self):
        for slider, label in self.playback_sliders:
            slider.deleteLater()
            label.deleteLater()
        self.playback_sliders.clear()

        y = 320
        font = QFont("ElysiaOSNew", 13)
        for client in self.get_playback_clients():
            label = QLabel(client.get("name", f"App {client['index']}"), self.sound_box)
            label.setFont(font)
            label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            label.move(10, y)

            slider = QSlider(Qt.Orientation.Horizontal, self.sound_box)
            slider.setGeometry(150, y, 400, 25)
            slider.setMinimum(0)
            slider.setMaximum(100)
            slider.setValue(client["volume"])
            slider.setStyleSheet(self.slider_style())
            slider.valueChanged.connect(lambda val, idx=client["index"]: self.run_cmd(f"pactl set-sink-input-volume {idx} {val}%"))
            slider.show()

            label.show()
            self.playback_sliders.append((slider, label))
            y += 40

    def show(self):
        self.audio_dropdown.clear()
        self.mic_dropdown.clear()

        for dev in self.get_devices("sink"):
            for port, desc in dev["ports"]:
                label = f"{dev['description']} - {desc}"
                self.audio_dropdown.addItem(label, (dev["name"], port))
                if dev["active"] == port:
                    self.audio_dropdown.setCurrentIndex(self.audio_dropdown.count() - 1)

        for dev in self.get_devices("source"):
            if ".monitor" in dev["name"]: continue
            for port, desc in dev["ports"]:
                label = f"{dev['description']} - {desc}"
                self.mic_dropdown.addItem(label, (dev["name"], port))
                if dev["active"] == port:
                    self.mic_dropdown.setCurrentIndex(self.mic_dropdown.count() - 1)

        self.refresh_state()
        self.sound_box.setVisible(True)
        self.parent.hide_tiles()
        self.parent.show_back_button()
        self.refresh_timer.start(1000)

    def hide(self):
        if self.audio_timer: self.audio_timer.stop()
        if self.mic_timer: self.mic_timer.stop()
        self.refresh_timer.stop()
        self.sound_box.setVisible(False)
        self.parent.hide_back_button()
        self.parent.show_tiles()

File: components/Sound.cpp
#include "Sound.h"
#include <QDir>
#include <QFont>
#include <QProcess>
#include <QVBoxLayout>
#include <QComboBox>
#include <QSlider>
#include <QPushButton>
#include <QLabel>
#include <QTimer>
#include <QMessageBox>
#include <QRegularExpression>
#include <QRegularExpressionMatch>
#include <QVariant>
#include <QMap>
#include <QPair>
#include <QTextEdit>
#include <QScrollArea>

Sound::Sound(QObject *parent) : QObject(parent) {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    audioTimer = nullptr;
    micTimer = nullptr;
    refreshTimer = new QTimer(this);
    connect(refreshTimer, &QTimer::timeout, this, &Sound::refreshState);
    initUI();
    refreshTimer->start(1000);
}

void Sound::initUI() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    soundBox = new QWidget(parentWidget);
    soundBox->setGeometry(500, 100, 600, 660);
    soundBox->setStyleSheet("background-color: rgba(255,255,255,0.05); border: 2px solid rgba(255,255,255,0.15); border-radius: 30px;");
    soundBox->setVisible(false);
    QFont font("ElysiaOSNew", 14);
    // Audio Output
    audioLabel = new QLabel("Audio Output", soundBox);
    audioLabel->setFont(font);
    audioLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    audioLabel->move(220, 30);
    audioDropdown = new QComboBox(soundBox);
    audioDropdown->setGeometry(150, 60, 300, 40);
    audioDropdown->setStyleSheet(
        "QComboBox {"
        "background: rgba(255, 255, 255, 0.2);"
        "color: white;"
        "border: 1px solid rgba(255, 255, 255, 0.2);"
        "border-radius: 36px;"
        "padding: 5px;"
        "font-size: 16px;"
        "}"
        "QComboBox QAbstractItemView {"
        "background: rgba(0, 0, 0, 0.2);"
        "border: 1px solid rgba(255, 255, 255, 0.2);"
        "color: black;"
        "padding: 5px;"
        "margin: 0px;"
        "selection-background-color: transparent;"
        "}"
    );
    connect(audioDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &Sound::setAudioOutput);
    audioSlider = new QSlider(Qt::Horizontal, soundBox);
    audioSlider->setGeometry(100, 110, 400, 30);
    audioSlider->setMinimum(0);
    audioSlider->setMaximum(100);
    audioSlider->setStyleSheet(
        "QSlider::horizontal { background: transparent; border: none; padding: 0px; margin: 0px; }"
        "QSlider::groove:horizontal { height: 8px; background: rgba(255, 255, 255, 0.1); border-radius: 4px; }"
        "QSlider::handle:horizontal { background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3); border: 1px solid white; width: 20px; margin: -8px 0; border-radius: 10px; }"
        "QSlider::sub-page:horizontal { background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3); border-radius: 4px; }"
    );
    connect(audioSlider, &QSlider::valueChanged, this, &Sound::debounceAudio);
    // Microphone Input
    micLabel = new QLabel("Microphone Input", soundBox);
    micLabel->setFont(font);
    micLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    micLabel->move(220, 170);
    micDropdown = new QComboBox(soundBox);
    micDropdown->setGeometry(150, 200, 300, 40);
    micDropdown->setStyleSheet(audioDropdown->styleSheet());
    connect(micDropdown, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &Sound::setMicInput);
    micSlider = new QSlider(Qt::Horizontal, soundBox);
    micSlider->setGeometry(100, 250, 400, 30);
    micSlider->setMinimum(0);
    micSlider->setMaximum(100);
    micSlider->setStyleSheet(audioSlider->styleSheet());
    connect(micSlider, &QSlider::valueChanged, this, &Sound::debounceMic);
    // Playback Clients
    clientsLabel = new QLabel("Apps Using Audio", soundBox);
    clientsLabel->setFont(font);
    clientsLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
    clientsLabel->move(220, 320);
    
    clientsScrollArea = new QScrollArea(soundBox);
    clientsScrollArea->setGeometry(100, 350, 400, 200);
    clientsScrollArea->setWidgetResizable(true);
    clientsScrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    clientsScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    
    clientsWidget = new QWidget();
    clientsLayout = new QVBoxLayout(clientsWidget);
    clientsLayout->setSpacing(15);
    clientsLayout->setContentsMargins(15, 15, 15, 15);
    clientsLayout->setAlignment(Qt::AlignTop);
    clientsScrollArea->setWidget(clientsWidget);
    // Apply transparent style after both widgets are constructed
    clientsScrollArea->setStyleSheet(
        "QScrollArea, QScrollArea > QWidget { color: white; background: transparent; border: none; margin: 0px; padding: 0px; }"
        "QScrollBar:vertical { background: rgba(255,255,255,0.1); width: 10px; border-radius: 5px; }"
        "QScrollBar::handle:vertical { background: rgba(229,167,198,0.6); border-radius: 5px; min-height: 20px; }"
        "QScrollBar::handle:vertical:hover { background: rgba(229,167,198,0.8); }"
    );
    clientsWidget->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px;");
    // Apply button
    applyBtn = new QPushButton("Apply Changes", soundBox);
    applyBtn->setGeometry(200, 570, 200, 50);
    applyBtn->setStyleSheet(
        "QPushButton {"
        "background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3);"
        "color: white;"
        "font-weight: bold;"
        "border: none;"
        "padding: 6px 12px;"
        "letter-spacing: 1px;"
        "border-radius: 10px;"
        "font-size: 16px;"
        "}"
        "QPushButton:hover {"
        "background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #edcee3, stop:1 #f5e5f0);"
        "}"
        "QPushButton:pressed {"
        "background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #d495b8, stop:1 #e5a7c6);"
        "}"
    );
    connect(applyBtn, &QPushButton::clicked, this, &Sound::applyChanges);
    updateDropdowns();
    updatePlaybackClients();
}

void Sound::show() {
    updateDropdowns();
    soundBox->setVisible(true);
}

void Sound::hide() {
    soundBox->setVisible(false);
}

QString simplifyDescription(const QString &text) {
    QString t = text;
    t.replace("Analog Stereo", "");
    QRegularExpression re1("Family\\s+[\\w/]+?\\s+HD\\s+Audio\\s+Controller", QRegularExpression::CaseInsensitiveOption);
    t.replace(re1, "");
    QRegularExpression re2("[\\s\\-]+$", QRegularExpression::CaseInsensitiveOption);
    t.replace(re2, "");
    return t.trimmed();
}

// Helper: get all sinks/sources with description, ports, and active port
struct AudioDevice {
    QString name;
    QString description;
    QList<QPair<QString, QString>> ports; // (port, port description)
    QString activePort;
};

QList<AudioDevice> getAudioDevices(const QString &type) {
    QList<AudioDevice> devices;
    QProcess shortProc, fullProc;
    shortProc.start("pactl", QStringList() << "list" << "short" << (type + "s"));
    shortProc.waitForFinished();
    QString shortOut = QString::fromUtf8(shortProc.readAllStandardOutput());
    fullProc.start("pactl", QStringList() << "list" << (type + "s"));
    fullProc.waitForFinished();
    QString fullOut = QString::fromUtf8(fullProc.readAllStandardOutput());
    QString blockHeader = type[0].toUpper() + type.mid(1) + " #";
    QStringList blocks = fullOut.split(blockHeader, Qt::SkipEmptyParts);
    QMap<QString, QString> nameToDesc;
    QMap<QString, QList<QPair<QString, QString>>> nameToPorts;
    QMap<QString, QString> nameToActivePort;
    for (const QString &block : blocks) {
        QString name, desc, activePort;
        QList<QPair<QString, QString>> ports;
        QStringList lines = block.split('\n');
        for (int i = 0; i < lines.size(); ++i) {
            QString line = lines[i].trimmed();
            if (line.startsWith("Name:")) {
                name = line.section(":", 1).trimmed();
            } else if (line.startsWith("Description:")) {
                desc = line.section(":", 1).trimmed();
            } else if (line.startsWith("Ports:")) {
                // Parse ports until next empty line or non-indented line
                int j = i + 1;
                while (j < lines.size() && (lines[j].startsWith(" ") || lines[j].startsWith("\t"))) {
                    QString portLine = lines[j].trimmed();
                    if (portLine.contains(":")) {
                        QString portName = portLine.section(":", 0, 0).trimmed();
                        QString portDesc = portLine.section(":", 1).section("(", 0, 0).trimmed();
                        ports.append(qMakePair(portName, portDesc));
                    }
                    ++j;
                }
            } else if (line.startsWith("Active Port:")) {
                activePort = line.section(":", 1).trimmed();
            }
        }
        if (!name.isEmpty()) {
            QString cleanDesc = desc.isEmpty() ? name : simplifyDescription(desc);
            nameToDesc[name] = cleanDesc;
            nameToPorts[name] = ports;
            nameToActivePort[name] = activePort;
        }
    }
    QStringList shortLines = shortOut.split('\n', Qt::SkipEmptyParts);
    for (const QString &line : shortLines) {
        QStringList parts = line.split('\t');
        if (parts.size() < 2) continue;
        QString devName = parts[1];
        QString desc = nameToDesc.value(devName, devName);
        QList<QPair<QString, QString>> ports = nameToPorts.value(devName);
        QString activePort = nameToActivePort.value(devName);
        devices.append({devName, desc, ports, activePort});
    }
    return devices;
}

void Sound::updateDropdowns() {
    // Audio Output
    audioDropdown->blockSignals(true);
    audioDropdown->clear();
    QList<AudioDevice> sinks = getAudioDevices("sink");
    int audioActiveIdx = 0, audioIdx = 0;
    for (const auto &dev : sinks) {
        for (const auto &port : dev.ports) {
            QString label = dev.description + " - " + port.second;
            audioDropdown->addItem(label, QVariant::fromValue(QPair<QString,QString>(dev.name, port.first)));
            if (dev.activePort == port.first) audioActiveIdx = audioIdx;
            ++audioIdx;
        }
    }
    audioDropdown->setCurrentIndex(audioActiveIdx);
    audioDropdown->blockSignals(false);
    // Mic Input
    micDropdown->blockSignals(true);
    micDropdown->clear();
    QList<AudioDevice> sources = getAudioDevices("source");
    int micActiveIdx = 0, micIdx = 0;
    for (const auto &dev : sources) {
        if (dev.name.contains(".monitor")) continue;
        for (const auto &port : dev.ports) {
            QString label = dev.description + " - " + port.second;
            micDropdown->addItem(label, QVariant::fromValue(QPair<QString,QString>(dev.name, port.first)));
            if (dev.activePort == port.first) micActiveIdx = micIdx;
            ++micIdx;
        }
    }
    micDropdown->setCurrentIndex(micActiveIdx);
    micDropdown->blockSignals(false);
}

QPair<QString,QString> getDropdownDevicePort(QComboBox *dropdown) {
    QVariant v = dropdown->currentData();
    if (v.canConvert<QPair<QString,QString>>()) {
        return v.value<QPair<QString,QString>>();
    }
    return QPair<QString,QString>();
}

void Sound::setAudioOutput() {
    auto pair = getDropdownDevicePort(audioDropdown);
    if (!pair.first.isEmpty() && !pair.second.isEmpty()) {
        QProcess::execute("pactl", QStringList() << "set-default-sink" << pair.first);
        QProcess::execute("pactl", QStringList() << "set-sink-port" << pair.first << pair.second);
    }
}

void Sound::setMicInput() {
    auto pair = getDropdownDevicePort(micDropdown);
    if (!pair.first.isEmpty() && !pair.second.isEmpty()) {
        QProcess::execute("pactl", QStringList() << "set-default-source" << pair.first);
        QProcess::execute("pactl", QStringList() << "set-source-port" << pair.first << pair.second);
    }
}

void Sound::setAudioVolume(int value) {
    auto pair = getDropdownDevicePort(audioDropdown);
    if (!pair.first.isEmpty()) {
        QProcess::execute("pactl", QStringList() << "set-sink-volume" << pair.first << QString::number(value) + "%");
    }
}

void Sound::setMicVolume(int value) {
    auto pair = getDropdownDevicePort(micDropdown);
    if (!pair.first.isEmpty()) {
        QProcess::execute("pactl", QStringList() << "set-source-volume" << pair.first << QString::number(value) + "%");
    }
}

void Sound::debounceAudio(int value) {
    if (audioTimer) audioTimer->stop();
    if (!audioTimer) audioTimer = new QTimer(this);
    audioTimer->setSingleShot(true);
    connect(audioTimer, &QTimer::timeout, this, [=]() { setAudioVolume(value); });
    audioTimer->start(100);
}

void Sound::debounceMic(int value) {
    if (micTimer) micTimer->stop();
    if (!micTimer) micTimer = new QTimer(this);
    micTimer->setSingleShot(true);
    connect(micTimer, &QTimer::timeout, this, [=]() { setMicVolume(value); });
    micTimer->start(100);
}

void Sound::applyChanges() {
    setAudioOutput();
    setMicInput();
    setAudioVolume(audioSlider->value());
    setMicVolume(micSlider->value());
    QMessageBox::information(soundBox, "Audio", "Settings applied.");
}

void Sound::refreshState() {
    updateSlider(audioDropdown, audioSlider, true);
    updateSlider(micDropdown, micSlider, false);
    updatePlaybackClients();
}

void Sound::updateSlider(QComboBox *dropdown, QSlider *slider, bool isSink) {
    auto pair = getDropdownDevicePort(dropdown);
    if (pair.first.isEmpty()) return;
    QString kind = isSink ? "sink" : "source";
    QProcess process;
    process.start("pactl", QStringList() << "get-" + kind + "-volume" << pair.first);
    process.waitForFinished();
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    QRegularExpression re("(\\d+%)");
    QRegularExpressionMatch match = re.match(output);
    if (match.hasMatch()) {
        QString percentStr = match.captured(0);
        int percent = percentStr.remove("%").toInt();
        if (slider->value() != percent) {
            slider->blockSignals(true);
            slider->setValue(percent);
            slider->blockSignals(false);
        }
    }
}

QList<QMap<QString, QString>> Sound::getPlaybackClients() {
    QProcess process;
    process.start("pactl", QStringList() << "list" << "sink-inputs");
    process.waitForFinished();
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    
    QList<QMap<QString, QString>> clients;
    QMap<QString, QString> current;
    
    QStringList lines = output.split('\n');
    for (const QString &line : lines) {
        QString trimmedLine = line.trimmed();
        if (trimmedLine.startsWith("Sink Input #")) {
            if (!current.isEmpty()) {
                clients.append(current);
            }
            current.clear();
            QString index = trimmedLine.split("#")[1];
            current["index"] = index;
        } else if (trimmedLine.startsWith("application.name")) {
            QRegularExpression re("\"([^\"]+)\"");
            QRegularExpressionMatch match = re.match(trimmedLine);
            if (match.hasMatch()) {
                current["name"] = match.captured(1);
            }
        } else if (trimmedLine.startsWith("Volume:")) {
            QRegularExpression re("(\\d+)%");
            QRegularExpressionMatchIterator it = re.globalMatch(trimmedLine);
            QList<int> volumes;
            while (it.hasNext()) {
                QRegularExpressionMatch match = it.next();
                volumes.append(match.captured(1).toInt());
            }
            if (!volumes.isEmpty()) {
                int avgVolume = 0;
                for (int vol : volumes) {
                    avgVolume += vol;
                }
                avgVolume /= volumes.size();
                current["volume"] = QString::number(avgVolume);
            }
        }
    }
    
    if (!current.isEmpty()) {
        clients.append(current);
    }
    
    return clients;
}

void Sound::updatePlaybackClients() {
    QList<QMap<QString, QString>> clients = getPlaybackClients();
    
    // Clear existing sliders and timers
    QLayoutItem *child;
    while ((child = clientsLayout->takeAt(0)) != nullptr) {
        delete child->widget();
        delete child;
    }
    
    // Clear app timers
    for (auto timer : appTimers.values()) {
        if (timer) timer->stop();
    }
    appTimers.clear();
    appSliders.clear();

    if (clients.isEmpty()) {
        QLabel *noAppsLabel = new QLabel("No applications are currently using audio.", clientsWidget);
        noAppsLabel->setAlignment(Qt::AlignCenter);
        noAppsLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        clientsLayout->addWidget(noAppsLabel);
    } else {
        for (const auto &client : clients) {
            QString name = client.value("name", "Unknown");
            QString volume = client.value("volume", "0");
            QString index = client.value("index", "");
            
            QWidget *appWidget = new QWidget();
            appWidget->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
            QVBoxLayout *appLayout = new QVBoxLayout(appWidget);
            appLayout->setSpacing(8);
            appLayout->setContentsMargins(0, 0, 0, 0);

            QLabel *appNameLabel = new QLabel(name);
            appNameLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px; font-weight: bold;");
            appLayout->addWidget(appNameLabel);

            QSlider *appVolumeSlider = new QSlider(Qt::Horizontal, appWidget);
            appVolumeSlider->setMinimum(0);
            appVolumeSlider->setMaximum(100);
            appVolumeSlider->setValue(volume.toInt());
            appVolumeSlider->setStyleSheet(
                "QSlider::horizontal { background: transparent; border: none; padding: 0px; margin: 0px; }"
                "QSlider::groove:horizontal { height: 6px; background: rgba(255, 255, 255, 0.1); border-radius: 3px; }"
                "QSlider::handle:horizontal { background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3); border: 1px solid white; width: 16px; margin: -6px 0; border-radius: 8px; }"
                "QSlider::sub-page:horizontal { background: qlineargradient(x1:0, y1:0, x2:1, y2:0, stop:0 #e5a7c6, stop:1 #edcee3); border-radius: 3px; }"
            );
            
            // Create debouncing timer for this app
            QTimer *appTimer = new QTimer(this);
            appTimer->setSingleShot(true);
            appTimer->setInterval(150); // Shorter debounce for better responsiveness
            appTimers[index] = appTimer;
            appSliders[index] = appVolumeSlider;
            
            connect(appVolumeSlider, &QSlider::valueChanged, this, [this, index, appVolumeSlider, appTimer]() {
                appTimer->stop();
                appTimer->disconnect();
                connect(appTimer, &QTimer::timeout, this, [this, index, appVolumeSlider]() {
                    QProcess::execute("pactl", QStringList() << "set-sink-input-volume" << index << QString::number(appVolumeSlider->value()) + "%");
                });
                appTimer->start();
            });
            
            appLayout->addWidget(appVolumeSlider);
            clientsLayout->addWidget(appWidget);
        }
    }
}

void Sound::onAppVolumeChanged(int value) {
    // This function is called when an app's volume slider changes
    // The actual volume change is handled in the lambda function above
    Q_UNUSED(value)
} 